src/query/query_visitor.rs

```diff
@@ -0,0 +1,225 @@
+        }
+        Mutation(inner) => {
+            visitor.visit_mutation(inner);
+            walk_mutation(visitor, inner);
+        }
+        Subscription(inner) => {
+            visitor.visit_subscription(inner);
+            walk_subscription(visitor, inner);
+        }
+    }
+}
+
+fn walk_query<'ast, V: QueryVisitor<'ast>>(visitor: &mut V, node: &'ast Query) {
+    visitor.visit_query(node);
+
+    for var_def in &node.variable_definitions {
+        visitor.visit_variable_definition(var_def);
+        walk_variable_definition(visitor, var_def);
+    }
+
+    visitor.visit_selection_set(&node.selection_set);
+    walk_selection_set(visitor, &node.selection_set);
+}
+
+fn walk_mutation<'ast, V: QueryVisitor<'ast>>(visitor: &mut V, node: &'ast Mutation) {
+    visitor.visit_mutation(node);
+
+    for var_def in &node.variable_definitions {
+        visitor.visit_variable_definition(var_def);
+        walk_variable_definition(visitor, var_def);
+    }
+
+    visitor.visit_selection_set(&node.selection_set);
+    walk_selection_set(visitor, &node.selection_set);
+}
+
+fn walk_subscription<'ast, V: QueryVisitor<'ast>>(visitor: &mut V, node: &'ast Subscription) {
+    visitor.visit_subscription(node);
+
+    for var_def in &node.variable_definitions {
+        visitor.visit_variable_definition(var_def);
+        walk_variable_definition(visitor, var_def);
+    }
+
+    visitor.visit_selection_set(&node.selection_set);
+    walk_selection_set(visitor, &node.selection_set);
+}
+
+fn walk_selection_set<'ast, V: QueryVisitor<'ast>>(visitor: &mut V, node: &'ast SelectionSet) {
+    visitor.visit_selection_set(node);
+
+    for selection in &node.items {
+        visitor.visit_selection(selection);
+        walk_selection(visitor, selection);
+    }
+}
+
+fn walk_variable_definition<'ast, V: QueryVisitor<'ast>>(visitor: &mut V, node: &'ast VariableDefinition) {
+    visitor.visit_variable_definition(node)
+}
+
+fn walk_selection<'ast, V: QueryVisitor<'ast>>(visitor: &mut V, node: &'ast Selection) {
+    use super::ast::Selection::*;
+
+    visitor.visit_selection(node);
+    match node {
+        Field(inner) => {
+            visitor.visit_field(inner);
+            walk_field(visitor, inner);
+        }
+        FragmentSpread(inner) => {
+            visitor.visit_fragment_spread(inner);
+            walk_fragment_spread(visitor, inner);
+        }
+        InlineFragment(inner) => {
+            visitor.visit_inline_fragment(inner);
+            walk_inline_fragment(visitor, inner);
+        }
+    }
+}
+
+fn walk_field<'ast, V: QueryVisitor<'ast>>(visitor: &mut V, node: &'ast Field) {
+    visitor.visit_field(node)
+}
+
+fn walk_fragment_spread<'ast, V: QueryVisitor<'ast>>(visitor: &mut V, node: &'ast FragmentSpread) {
+    visitor.visit_fragment_spread(node)
+}
+
+fn walk_inline_fragment<'ast, V: QueryVisitor<'ast>>(visitor: &mut V, node: &'ast InlineFragment) {
+    visitor.visit_inline_fragment(node);
+}
```

Comment:
Same as for fragments, would it make sense to visit the selection set?

Reply:
Addressed in cb71a7e 👍